README.md
# vending-machine-business

This repository will run a simulation with several different vending machines in differente locations with different id's

## Installation

Download from http://example.com/FIXME.

## Usage

To run the executable found in target use the following commands

	$ java -jar vending-machine-business-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.
# vending-machine-business


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

### TODO
- Make the program accept a file via command line interface and read from it.
- Set a variable to be modified at runtime.
- Execute the transaction to the data
- Write data to the same file

### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

### Added


test/vending_machine_business/core_test.clj
(ns vending-machine-business.core-test
  (:require [clojure.test :refer :all]
            [vending-machine-business.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/vending_machine_business/helper.clj
(ns vending-machine-business.helper)
(use 'clojure.java.io)

(defn usage []
  (println "Usage: vending-business [OPTIONS] ... [FILE] ...")
  (println)
  (println "Simulation of several vending machines. To run, pass a file with the vending machines.")
  (println "The format for the file will be ((nombre-maquina(('nombre-producto' precio cantidad) ...)((valor disponibilidad) ... )('nombre-producto' (moneda moneda moneda))...) ... )")
  (println)
  (println "Options")
  (println "-f, --file FILE                 specify file with database.")
  (println "-s, --simulation                will run a simulation with default data and will save it in db/ project folder.")
  (println "-m, --machine MACHINE           will select a machine to work with individually. ")
  (println "-t, --transaction TRANSACTION   will use a transaction file with format transaction::- (('nombre producto' (moneda ...)) ...) ")

  (println)
  (println "-h, --help                      shows this help menu and exit.")
  (println "-v, --version                   displays current version of script.\n")
)

(defn write-file [file-path data]
  (spit file-path
        (pr-str data)
  )
)

(defn read-file [file-path]
  (read-string (slurp file-path))
)

;; Define paths to save the data
(def business-path "/home/neto/Documents/Tec/4thSemester/IMC/EvidenciaTres/vending-machine-business/src/vending_machine_business/business-data")
(def coin-path "/home/neto/Documents/Tec/4thSemester/IMC/EvidenciaTres/vending-machine-business/src/vending_machine_business/coin-won")

;; Default machines
(def default-machine
  '(
    ("first-machine" (("coca" 15 10)("agua" 15 12)("manzanita" 7 10)) ((10 20)(5 30)(2 15)(1 20)) (("coca" (1 1 1 5 10))("chocoroles" (1 1 1 5 10))("agua" (1 1 1 5 10))))
    ("second-machine" (("gansito" 18 6) ("pinguinos" 15 17) ("coca" 20 20)) ((10 20)(5 22)(2 15)(1 50)) (("coca" (1 1 1 5 10))("chocoroles" (1 1 1 5 10))("agua" (1 1 1 5 10))))
    ("third-machine" (("manzanita" 20 20) ("agua" 15 20) ("chocoroles" 18 20)) ((10 20)(5 31)(2 11)(1 12)) (("coca" (1 1 1 5 10))("chocoroles" (1 1 1 5 10))("agua" (1 1 1 5 10))))
    ("fourth-machine" (("fresca" 18 17) ("fritos" 15 17) ("ruffles" 15 20)) ((10 30)(5 5)(2 5)(1 31)) (("coca" (1 1 1 5 10))("chocoroles" (1 1 1 5 10))("agua" (1 1 1 5 10))))
    ("fifth-machine" (("coca" 15 10)("fresca" 18 5)("manzanita" 10 23)) ((10 53)(5 20)(2 10)(1 35)) (("coca" (1 1 1 5 10))("chocoroles" (1 1 1 5 10))("agua" (1 1 1 5 10))))
    ("sixth-machine" (("coca" 15 15) ("pinguinos" 15 7) ("ruffles" 15 14)) ((20 23)(10 10)(5 10)(2 40)(1 26))(("coca" (1 1 1 5 10))("chocoroles" (1 1 1 5 10))("agua" (1 1 1 5 10))))
  )
)

(def coin-won ;; a unmutable data container to represent the information gain
  '(
    ("first-machine"((50 0)(20 0)(10 0)(5 0)(2 0)(1 0)))
    ("second-machine"((50 0)(20 0)(10 0)(5 0)(2 0)(1 0)))
    ("third-machine"((50 0)(20 0)(10 0)(5 0)(2 0)(1 0)))
    ("fourth-machine"((50 0)(20 0)(10 0)(5 0)(2 0)(1 0)))
    ("fifth-machine"((50 0)(20 0)(10 0)(5 0)(2 0)(1 0)))
    ("sixth-machine"((50 0)(20 0)(10 0)(5 0)(2 0)(1 0)))
  )
)

(defn create-coin-won []
  (write-file coin-path coin-won)
)

(defn create-default-machine []
  (write-file business-path default-machine)
)

(defn read-business-data []
  (read-file business-path)
)

(defn read-won-coin []
 (read-file coin-path)
)

(def max-deposit 50)

;;
;; helper.rkt defines the initial data for the simulation of the machine and helper functions to be used through the rest of the program
;;

(defn reconstruct-deposit [datos index value]
    (cond
     (nil? datos) '()
     (= index (first (first datos))) (concat (list (list (first (first datos)) (+ (first (rest (first datos))) value))) (rest datos))
     :else (concat (list (first datos)) (reconstruct-deposit (rest datos) index value))
    )
)

(defn reconstruct-inventory [datos index value]
      (cond
       (nil? datos) '()
       (= index (first (first datos))) (concat (list (list (first (first datos)) (first (rest (first datos))) (+ (first (rest (rest (first datos)))) value))) (rest datos))
       :else (concat (list (first datos)) (reconstruct-inventory (rest @datos) index value))
      )
)

 (defn update-coin-won [index value]
  (println "Updating coin won")
  () ;; open won coin
  (println coin-won)
  ;; save coin-won in a file to be opened next time
 )

(defn update-money [index value inventory]
  ; (set! deposit (reconstruct-deposit deposit index value))
  (println "Setting deposit")
)

(defn update-inventory [index value inventory]
  ; (set! inventory (reconstruct-inventory inventory index value))
  (println "Updating inventory")
)


;; Set and get data to use in the main code:
;; Make copies of the inventory and deposit
;; Update stuff
;; Retrieve data is needed

(defn bubble [lts] ;; bubble sort for change automata
  (if (nil? (rest lts))
      lts
      (if (> (first (first lts)) (first (first (rest lts))))
          (cons (first lts)
                (bubble (rest lts)))
          (cons (first (rest lts))
                (bubble (cons (first lts) (rest (rest lts)))))
      )
  )
)

(defn bubble-sort [N lts]
  (cond
        (= N 1) (bubble lts)
        :else (bubble-sort (- N 1) (bubble lts))
  )
)

(defn set-bubble [lts]
  (bubble-sort (.length lts) lts)
)

;; Check data from the money deposit
(defn get-coin [index coin-list]
  (cond
    (= index (first (first coin-list))) (first (rest (first coin-list)))
    :else (get-coin index (rest coin-list))
  )
)

(defn get-space [index deposit]
  (get-coin index deposit)
)


;; Check data from the inventory
(defn get-product [index inv]
  (cond
    (= index (first (first inv))) (first (rest (rest (first inv))))
    :else (get-product index (rest inv))
  )
)

(defn get-available [index inventory]
  (get-product index inventory)
)

; ;; Print inventory info. If get-available < 10 then is almost empty
; (defn almost-empty-inventory [inv]
;   (cond
;     (nil? inv) true
;     (< (get-available (first (first inv))) 10) (println (first (first inv))) (println " is almost empty or empty") (almost-empty-inventory (rest inv)) ;; if almost full
;   )
; )

; ;; Print deposit info. If get-space is less than 10, then almos empty. If bigger than 50, empty
; (defn almost-full-coin []
;   (cond
;     (> (get-space 1) (- max-deposit 10)) (println "The deposit for coins of 1 is almost full or full, with: ") (println (get-space 1))  ;; if almost full
;   )
;   (cond
;     (> (get-space 2) (- max-deposit 10)) (println "The deposit for coins of 2 is almost full or full, with: ") (println (get-space 2)) ;; if almost full
;   )
;   (cond
;     [(> (get-space 5) (- max-deposit 10)) (println "The deposit for coins of 5 is almost full or full, with: ") (println (get-space 5)) ] ;; if almost full
;   )
;   (cond
;     [(> (get-space 10) (- max-deposit 10)) (println "The deposit for coins of 10 is almost full or full, with: ") (println (get-space 10)) ] ;; if almost full
;   )
;   (cond
;     [(> (get-space 20) (- max-deposit 10)) (println "The deposit for coins of 20 is almost full or full, with: ") (println (get-space 20)) ] ;; if almost full
;   )
;   (cond
;     [(> (get-space 50) (- max-deposit 10)) (println "The deposit for coins of 50 is almost full or full, with: ") (println (get-space 50)) ] ;; if almost full
;   )
; )

; (defn almost-empty-coin []
;   (cond
;     (< (get-space 1) 10) (println "The deposit for coins of 1 is almost empty or empty, with: ") (println (get-space 1))  ;; if almost empty
;   )
;   (cond
;     (< (get-space 2) 10) (println "The deposit for coins of 2 is almost empty or empty, with: ") (println (get-space 2))  ;; if almost empty
;   )
;   (cond
;     (< (get-space 5) 10) (println "The deposit for coins of 5 is almost empty or empty, with: ") (println (get-space 5))  ;; if almost empty
;   )
;   (cond
;     (< (get-space 10) 10) (println "The deposit for coins of 10 is almost empty or empty, with: ") (println (get-space 10))  ;; if almost empty
;   )
;   (cond
;     (< (get-space 20) 10) (println "The deposit for coins of 20 is almost empty or empty, with: ") (println (get-space 20))  ;; if almost empty
;   )
;   (cond
;     (< (get-space 50) 10) (println "The deposit for coins of 50 is almost empty or empty, with: ") (println (get-space 50))  ;; if almost empty
;   )
; )


src/vending_machine_business/core.clj
(ns vending-machine-business.core
  (:gen-class)
)

(require '[vending-machine-business.helper :as hp])

;; Return the coint to user. I.E erase coin from deposit
(defn return-coin [coin]
  (hp/update-money coin -1)
  (hp/update-coin-won coin -1)
)

(defn fare-automata [deposit debt] ;; in this automata the deposit has to be checked to decide what to return
  ;; debt is input - cost what the machine owes to the user
  ;; debt = (- current-state final-state)

  (println)
  (println "Deposit in fare-automata " deposit)
  (println "Debt in fare-automata " debt)
  (println)

  (cond
    (= debt 0) true ;; [updates the money and inventory db]
                    ;; first(first money-deposit i)s coin value to update in db
                    ;; first (rest (first money-deposit i))s existance

    ;; If debt > 0 then select the first coin that is less or equal to deb
    ;; insert the coin
    (> debt 0)
     (cond
       (nil? deposit) false
       (or (and (< (first(first deposit)) debt) (> (first (rest (first deposit))) 0)) (and (= (first(first deposit)) debt) (> (first (rest (first deposit))) 0))) (do (return-coin (first(first deposit))) (fare-automata deposit (- debt (first(first deposit)))))
       :else (fare-automata (rest deposit) debt)
     )

    :else false ;; does not update and returns the security-data to the document
  )
)

(defn check-currency-in [transactions deposit inventory]
  ; (println "Transaction in check-currency-in: " transactions)
  (println "Transaction in check-currency-in:" (first transactions))
  (println "Inventory in check-currency-in :" inventory)

  (cond
    (nil? transactions) true
    (< (hp/get-space (first transactions) deposit) hp/max-deposit) (do (hp/update-coin-won (first transactions) 1 inventory deposit) (hp/update-money (first transactions) 1 inventory deposit) (check-currency-in (rest transactions) deposit inventory))
    ; (and (< (hp/get-space 1 deposit) hp/max-deposit) (= (first transactions) 1)) (do (hp/update-coin-won 1 1) (hp/update-money (first transactions) 1) (check-currency-in (rest transactions) deposit))
    ; (and (< (hp/get-space 2 deposit) hp/max-deposit) (= (first ) 2)) (do(hp/update-coin-won 2 1) (hp/update-money (first transactions) 1) (check-currency-in (rest transactions) deposit))
    ; (and (< (hp/get-space 5 deposit) hp/max-deposit) (= (first transactions) 5))  (do(hp/update-coin-won 5 1)(hp/update-money (first transactions) 1) (check-currency-in (rest transactions) deposit))
    ; (and (< (hp/get-space 10 deposit) hp/max-deposit) (= (first transactions) 10)) (do(hp/update-coin-won 10 1)(hp/update-money (first transactions) 1) (check-currency-in (rest transactions) deposit))
    ; (and (< (hp/get-space 20 deposit) hp/max-deposit) (= (first transactions) 20)) (do(hp/update-coin-won 20 1)(hp/update-money (first transactions) 1) (check-currency-in (rest transactions) deposit))
    ; (and (< (hp/get-space 50 deposit) hp/max-deposit) (= (first transactions) 50)) (do(hp/update-coin-won 50 1)(hp/update-money (first transactions) 1) (check-currency-in (rest transactions) deposit))

    :else false
  )
)

;; HERE DATA SHOULD BE UPDATED
;; current-state = money deposited by user
;; final-state = money that user should deposit
(defn validate [transactions final-state current-state deposit]

  (println)
  (println "Final-state in validate " final-state)
  (println "Current-state in validate " final-state)
  (println "Deposit in validate " deposit)
  (println "Transaction in validate " transactions)
  (println)

  (cond
    ;; If transactions not empty yet, then continue to add up the transactions inside the function
    (not(nil? transactions)) (validate (rest transactions) final-state (+ current-state (first transactions)) )

    ;; validated transactions without fare to return. update inv and money using transactions data
    (= final-state current-state) true

    ;; Change automata only needs to substract the expected value of the product to the value inputed by the user
    ;; For example, if a product costs 50 and user gave 55, then the machine must return 5 (input - cost)
    (> current-state final-state) (fare-automata deposit (- current-state final-state)) ;; starts change automata and updates stuff

    (< current-state final-state) false ;; transaction does not buy a product and db's not updated
  )
)

;; [compute-transaction] validates the transaction and updates de db
;; Getting money in
(defn compute-transaction [inventory transaction deposit]
  ;; (first (rest transaction)) is the list of transitions form '(1 1 2 5 10 20 50)
  ;; (first (rest inventory)) is the price form '("product-name" price quantity) : ("gansitos" 17 2)

  (println)
  (println "Inventory in compute-transaction: " inventory)
  (println "Deposit in compute-transaction: " deposit)
  (println "Transaction in compute-transaction: " transaction)
  (println)

  (cond
    (and (check-currency-in (first (rest transaction)) deposit inventory) (validate (first (rest transaction)) (first (rest inventory)) 0 deposit)) true
    :else false
  )
)

;; helper function of [start-transaction]
;; if product-name exist then return true
(defn product-exist? [inventory transaction deposit]

  (println "Inventory received in product-exist?: " inventory)
  (println "Deposit received in product-exist?: " deposit)
  (println "Transaction received in product-exist?: " transaction)

  (cond
    ;; if inventory null then all inventory checked, if compute-transaction is false then the transaction is not valid, so returns false
    (nil? inventory) false
    (and (and (= (first(first inventory)) (first transaction)) (> ( first (rest (rest (first inventory)))) 0)) (compute-transaction (first inventory) transaction deposit)) (do (hp/update-inventory (first(first inventory)) -1) true)
    :else (product-exist? (rest inventory) transaction deposit)
  )
)

(defn start-transactions [inventory deposit transactions]
  ;; get product in inventory and confirm it exist
  (println "Inventory received: " inventory)
  (println "Deposit received: " deposit)
  (println "Transactions received: " transactions)

  (cond
    (nil? transactions) true
    (product-exist? inventory (first transactions) deposit) (start-transactions inventory (rest transactions)) ;; executes the transaction
    :else false
  )
)

(defn start-transaction [machine]
  ;; get product in inventory and confirm it exist
  (cond
    (product-exist? inventory transaction deposit) true ;; executes the transaction
    :else false
  )
)


(defn final-stuff []
  (println "Final stuff")
)

(defn pass-list-to-automata [data] ;; this is the wrapper for the default data

  (println "Data received:" data)
  (println)

  (println "Machine name: " (first(first data)) )
  (println "Inventory: " (first (rest (first data))))
  (println "Deposit: " (first (rest (rest (first data)))))
  (println "Transaction: "(first (rest (rest (rest (first data))))))

  (cond
    (nil? data) (final-stuff) ;; if data is null then no more data found
    (= (start-transactions (first data) true) (println "Transaction completed");; inventory
  )
)

(defn pass-file-to-automata [data] ;; this is the wrapper for the specific data
  (println "Final stuff")
)

(defprotocol Format
  (fmt [this])
)
(extend-protocol Format
  clojure.lang.IPersistentList
    (fmt [this] (str "list"))
  clojure.lang.ArraySeq
    (fmt [this] (str "file"))
  clojure.lang.Atom
    (fmt [this] (str "atom"))
)

(defn run-machines [file]
  (println "Reaching this function")
  (cond
    (= (fmt file) "list") (pass-list-to-automata file)
    (= (fmt file) "file") (pass-file-to-automata file)
    :else (println "Cannot process the file.")
  )
)

;; -f / -s mutually exclusive.
;; -h / -v mutually exclusive
;; -m has to be run with -t

(defn wrapper-for-sim []
  (println "::- [ Business not specified. Creating default business ]")
  (hp/create-default-machine)
  (hp/create-coin-won)
  (run-machines (hp/read-business-data))
)


(defn validate-args [args]
  (println args) ;; access args data with first
  (cond
    (or (= (first args) "-f") (= (first args) "--file")) (run-machines (rest args))
    (or (= (first args) "-s") (= (first args) "--simulation")) (wrapper-for-sim)
    :else (println "Error. Ending the program.")
  )
)


(defn -main
  "Here I can provide a brief description of each function."
  [& args]
  (cond
    (nil? args) (hp/usage)
    :else (validate-args args)
  )
)
